src/chapter_3/PDAs.md

```diff
@@ -0,0 +1,315 @@
+#[program]
+pub mod game {
+    use super::*;
+    // handler function
+    pub fn create_user_stats(ctx: Context<CreateUserStats>, name: String) -> ProgramResult {
+        let user_stats = &mut ctx.accounts.user_stats;
+        user_stats.level = 0;
+        if name.as_bytes().len() > 200 {
+            // proper error handling omitted for brevity
+            panic!();
+        }
+        user_stats.name = name;
+        user_stats.bump = *ctx.bumps.get("user_stats").unwrap();
+        Ok(())
+    }
+}
+
+#[account]
+pub struct UserStats {
+    level: u16,
+    name: String,
+    bump: u8,
+}
+
+// validation struct
+#[derive(Accounts)]
+pub struct CreateUserStats<'info> {
+    #[account(mut)]
+    pub user: Signer<'info>,
+    // space: 8 discriminator + 2 level + 4 name length + 200 name + 1 bump
+    #[account(
+    init,
+    payer = user,
+    space = 8 + 2 + 4 + 200 + 1, seeds = [b"user-stats", user.key().as_ref()], bump)]
+    pub user_stats: Account<'info, UserStats>,
+    pub system_program: Program<'info, System>,
+}
+```
+
+In the account validation struct we use `seeds` together with `init` to create a PDA with the desired seeds.
+Additionally, we add an empty `bump` constraint to signal to anchor that it should find the bump itself.
+Then, in the handler, we call `ctx.bumps.get("user_stats")` to get the bump anchor found and save it to the user stats
+account as an extra property.
+
+If we then want to use the created pda in a different instruction, we can do the following:
+```rust,ignore
+// validation struct
+#[derive(Accounts)]
+pub struct ChangeUserName<'info> {
+    pub user: Signer<'info>,
+    #[account(mut, seeds = [b"user-stats", user.key().as_ref()], bump = user_stats.bump)]
+    pub user_stats: Account<'info, UserStats>,
+}
+
+// handler function
+pub fn change_user_name(ctx: Context<ChangeUserName>, new_name: String) -> ProgramResult {
+    if new_name.as_bytes().len() > 200 {
+        // proper error handling omitted for brevity
+        panic!();
+    }
+    ctx.accounts.user_stats.name = new_name;
+    Ok(())
+}
+```
+
+This will check that the `user_stats` account is the pda created by running `hash(seeds, user_stats.bump, game_program_id)`.
+
+Finally, let's add a test. Copy this into `game.ts`
+
+```ts
+import * as anchor from '@project-serum/anchor';
+import { web3 } from '@project-serum/anchor';
+import { Program } from '@project-serum/anchor';
+import { Game } from '../target/types/game';
+import { expect } from 'chai';
+
+describe('game', async() => {
+
+anchor.setProvider(anchor.Provider.env());
+
+  const program = anchor.workspace.Game as Program<Game>;
+
+  it('Sets and changes name!', async () => {
+    const [userStatsPDA, _] = await web3.PublicKey
+      .findProgramAddress(
+        [
+          anchor.utils.bytes.utf8.encode("user-stats"),
+          anchor.getProvider().wallet.publicKey.toBuffer()
+        ],
+        program.programId
+      );
+
+    await program.rpc.createUserStats("brian", {
```

Comment:
We should use the `methods` namespace throughout the book.

Reply:
yes. Although that also requires changing previous chapters and imo would be better done in a separate PR.